File: solutions/rust/poker/2/src/hands/poker_hand.rs
use std::collections::HashMap;
use std::collections::HashSet;

use lazy_static::lazy_static;

use crate::hands::poker_hand_value::PokerHandValue;
use crate::models::poker_card::PokerCard;
use crate::utilities::length_check::LengthCheck;

pub struct PokerHand {
	pub(crate) cards: Vec<PokerCard>,
}

lazy_static! {
    static ref RANK_GROUP_EVALUATIONS: [(fn(&HashMap<String, Vec<&PokerCard>>) -> bool, PokerHandValue); 5] = [
	((|groups: &HashMap<String, Vec<&PokerCard>>| groups.has_group_with_length(4)) as fn(&HashMap<String, Vec<&PokerCard>>) -> bool, PokerHandValue::FourOfAKind),
	((|groups: &HashMap<String, Vec<&PokerCard>>| groups.has_group_with_length(3) && groups.has_group_with_length(2)) as fn(&HashMap<String, Vec<&PokerCard>>) -> bool, PokerHandValue::FullHouse),
	((|groups: &HashMap<String, Vec<&PokerCard>>| groups.has_group_with_length(3)) as fn(&HashMap<String, Vec<&PokerCard>>) -> bool, PokerHandValue::ThreeOfAKind),
	((|groups: &HashMap<String, Vec<&PokerCard>>| groups.has_two_pair()) as fn(&HashMap<String, Vec<&PokerCard>>) -> bool, PokerHandValue::TwoPair),
	((|groups: &HashMap<String, Vec<&PokerCard>>| groups.has_group_with_length(2)) as fn(&HashMap<String, Vec<&PokerCard>>) -> bool, PokerHandValue::Pair),
];

static ref  HAND_CONDITION_EVALUATIONS: [(fn(&PokerHand) -> bool, PokerHandValue); 5] = [
	((|poker_hand: &PokerHand| poker_hand.is_flush() && poker_hand.is_low_ace_straight()) as fn(&PokerHand) -> bool, PokerHandValue::LowAceStraightFlush),
	((|poker_hand: &PokerHand| poker_hand.is_flush() && poker_hand.is_straight()) as fn(&PokerHand) -> bool, PokerHandValue::StraightFlush),
	((|poker_hand: &PokerHand| poker_hand.is_flush()) as fn(&PokerHand) -> bool, PokerHandValue::Flush),
	((|poker_hand: &PokerHand| poker_hand.is_low_ace_straight()) as fn(&PokerHand) -> bool, PokerHandValue::LowAceStraight),
	((|poker_hand: &PokerHand| poker_hand.is_straight()) as fn(&PokerHand) -> bool, PokerHandValue::Straight),
];
}

impl PokerHand {
	pub fn new(cards: Vec<PokerCard>)
	           -> Self {
		Self {
			cards
		}
	}


	pub fn get_hand_value(&self)
	                      -> PokerHandValue {
		Self::evaluate_hand_value(self)
	}

	fn evaluate_hand_value(&self) -> PokerHandValue {
		let rank_groups = self.create_rank_groups();

		for &(eval_fn, ref value) in RANK_GROUP_EVALUATIONS.iter() {
			if eval_fn(&rank_groups) {
				return *value; // Dereference value to get the actual PokerHandValue
			}
		}

		for &(eval_fn, ref value) in HAND_CONDITION_EVALUATIONS.iter() {
			if eval_fn(self) {
				return *value; // Dereference value to get the actual PokerHandValue
			}
		}

		PokerHandValue::HighCard
	}

	fn create_rank_groups(&self)
	                      -> HashMap<String, Vec<&PokerCard>> {
		let mut rank_groups: HashMap<String, Vec<&PokerCard>> = HashMap::new();
		for card in &self.cards {
			rank_groups.entry(card.rank.clone()).or_insert_with(Vec::new).push(&card);
		}
		rank_groups
	}

	fn is_straight(&self)
	               -> bool {
		let mut sorted_values: Vec<i32> = self.cards.iter().map(|card| card.value()).collect();
		sorted_values.sort_unstable();
		sorted_values.windows(2).all(|w| w[0] + 1 == w[1])
	}

	fn is_low_ace_straight(&self)
	                       -> bool {
		let mut sorted_ranks: Vec<String> = self.cards.iter().map(|card| card.rank.clone()).collect();
		sorted_ranks.sort_unstable();
		sorted_ranks == vec!["2", "3", "4", "5", "A"]
	}

	fn is_flush(&self)
	            -> bool {
		let suit_set: HashSet<String> = self.cards.iter().map(|card| card.suit.clone()).collect();
		suit_set.len() == 1
	}
}

File: solutions/java/poker/1/src/main/java/PokerHand.java
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class PokerHand {
    public PokerHand(Collection<PokerCard> cards) {
        if (cards == null || cards.isEmpty()) {
            throw new IllegalArgumentException("Cards must not be null or empty");
        }
        this.cards = cards;
    }

    private final Collection<PokerCard> cards;
    public Collection<PokerCard> getCards() {
        return this.cards;
    }

    private PokerHandValue handValue;
    public PokerHandValue getHandValue() {
        if (this.handValue == null) {
            this.handValue = this.evaluateHandValue();
        }
        return this.handValue;
    }

    private PokerHandValue evaluateHandValue() {
        Map<String, List<PokerCard>> rankGroups = this.cards
                .stream()
                .collect(Collectors.groupingBy(PokerCard::rank));
        Collection<List<PokerCard>> groups = rankGroups.values();
        boolean hasFourOfKind = groups.stream()
                                      .anyMatch(g -> g.size() == 4);
        if (hasFourOfKind) {
            return PokerHandValue.FOUR_OF_A_KIND;
        }

        boolean hasThreeOfKind = groups.stream()
                                       .anyMatch(g -> g.size() == 3);
        boolean hasPair = groups.stream()
                                .anyMatch(g -> g.size() == 2);
        long pairCount = groups.stream()
                               .filter(g -> g.size() == 2)
                               .count();
        if (hasThreeOfKind) {
            return hasPair
                   ? PokerHandValue.FULL_HOUSE
                   : PokerHandValue.THREE_OF_A_KIND;
        }

        if (pairCount == 2) {
            return PokerHandValue.TWO_PAIR;
        }

        if (hasPair) {
            return PokerHandValue.PAIR;
        }

        boolean flush = this.isFlush();
        boolean lowAceStraight = this.isLowAceStraight();
        boolean straight = this.isStraight();
        return flush
               ? lowAceStraight
                     ? PokerHandValue.LOW_ACE_STRAIGHT_FLUSH
                     : straight
                           ? PokerHandValue.STRAIGHT_FLUSH
                           : PokerHandValue.FLUSH
               : lowAceStraight
                     ? PokerHandValue.LOW_ACE_STRAIGHT
                     : straight
                           ? PokerHandValue.STRAIGHT
                           : PokerHandValue.HIGH_CARD;
    }

    private boolean isStraight() {
        List<Integer> sortedValues =
                cards.stream()
                     .map(PokerCard::getValue)
                     .sorted()
                     .toList();
        int start = sortedValues.get(0);
        return IntStream.range(start, start + 5)
                        .boxed()
                        .toList()
                        .equals(sortedValues);
    }

    private boolean isLowAceStraight() {
        return cards.stream()
                    .map(PokerCard::rank)
                    .sorted()
                    .toList()
                    .equals(List.of("2", "3", "4", "5", "A"));
    }

    private boolean isFlush() {
        String suit = cards.iterator()
                           .next()
                           .suit();
        return cards.stream()
                    .allMatch(card -> card.suit().equals(suit));
    }
}